Thecarisma-syntax.md

```diff
@@ -0,0 +1,620 @@
+    ```
+
+#### Why UFCS is useful
+
+Sometimes it makes sense to have certain functions be instance methods because they associate closely to a particular type. Other times it makes sense for such functions to be regular top-level functions because they don't associate with any single type. With UFCS, you don't have to memorize which one is a function or which one is an instance method, because you can call them using the both dot notation call and regular call syntaxes.
+
+
+- Take `map` function for example. It is a generic function and does not associate with any single type.
+
+    ```swift
+    func map[T, U](iter: T, fn: (U) -> U) -> T {
+        var new = T::new()
+        for (i, x) in enumerate(iter) {
+            new[i] = fn(x)
+        }
+        new
+    }
+    ```
+
+    Because we have UFCS support, we can chain up calls to regular function using the dot notation syntax, which can be easier to read sometimes.
+
+    ```swift
+    map(filter(map(arr, (x) => x * 2), (x) => x > 2), (x) => x * 2)
+    ```
+
+    ```swift
+    arr
+        .map((x) => x * 2)
+        .filter((x) => x > 2)
+        .map((x) => x * 2)
+    ```
+
+
+
+## TRAITS
+- Ratio does not support inheritance. Instead it choses composition-based type extension.
+
+    ```rust
+    trait Callable {
+        call: (...Int) -> Int
+    }
+    ```
+
+    ```swift
+    impl Sum extends Callable {
+        func call(args: ...Int) -> Int {
+            var result = 0
+            for i in args {
+                result += i
+            }
+            result
+        }
+    }
+    ```
+
+## UNSAFE BLOCK
+- The unsafe block is necessary for scoping the usage of unsafe/unstable features like pointer arithmetic and destructive casting.
+
+    ```rust
+    const base_ptr = some_ref as *Int
+
+    const base_ptr = unsafe {
+        base_ptr.add(20)
+    }
+    ```
+
+
+## CASTING
+- Non-destructive casting uses `as` keyword and it is only supported for scalar types.
+
+    ```rust
+    const val = 1000 as f32
+    ```
+
+## ACCESS MODIFIERS
+- Everything is private by default, but you can make member of a module or type public using the `public` keyword.
+
+    ```swift
+    public type Person {
+        pub name,
+        age
+    }
+    ```
+
+    ```swift
+    public func new(name, age) -> self {
+        self {
+            name, age
+        }
+    }
+    ```
+
+
+## IMPORT
+- Ratio is a module-based language. To use things from other module, you have to import from it.
+
+    ```py
+    import module
+
+    module.foo()
+    ```
+
+    ```py
+    import path.to.some.module {foo}
+    import module {foo as bar}
+
+    foo()
+    bar()
+    ```
+
+- Ratio does not allow cyclic imports/dependencies.
+
+## GENERICS
+- Declaring a generic type
+
+    ```rust
+    type Buffer:[T] {
+        buffer: *T,
+        capacity,
+        len
+    }
+    ```
+
+- Declaring a generic function
+
+    ```swift
+    impl Buffer:[T] where T < Integer {
```

Comment:
There's quite a lot going on here. Apparently, we are enforcing that the size of `T` is less than an `Integer`.

What operators are permitted in such scenarios?

I'll advise we give this a thoughtful and close look before proceeding.

Reply:
This syntax carries a lot of implementation implications going forward. We really need to get much more grounded on what implementation will look like in time.